File: server/src/processor_output_policy.rs
//! Automatic output capacity and input staging for processing workshops.
//!
//! Workshops produce up to physical capacity and stage three complete input
//! cycles. The legacy percentage column remains in stored rows for additive
//! schema compatibility, but no longer changes workshop behavior.

pub const PROCESSOR_OUTPUT_TARGET_DEFAULT_PERCENT: u8 = 100;
pub const PROCESSOR_INPUT_STAGING_DEFAULT_CYCLES: f64 = 3.0;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ProcessorOutputKind {
    Flour,
    Food,
    Ale,
    PreservedFood,
    TextileIntermediate,
    Cloth,
    Charcoal,
    Ironwork,
    Pottery,
    Leather,
    Shoes,
    Candles,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ProcessorInputKind {
    Grain,
    Flour,
    Water,
    Firewood,
    Barley,
    Food,
    Salt,
    Wool,
    Flax,
    Yarn,
    Linen,
    Iron,
    Charcoal,
    Clay,
    Apples,
    Honey,
    Hides,
    Leather,
    Wax,
}

pub fn processor_output_kind(kind: &str) -> Option<ProcessorOutputKind> {
    match kind {
        "watermill" | "windmill" => Some(ProcessorOutputKind::Flour),
        "bakery" => Some(ProcessorOutputKind::Food),
        "brewery" => Some(ProcessorOutputKind::Ale),
        "smokehouse" => Some(ProcessorOutputKind::PreservedFood),
        "spinning_retting_house" => Some(ProcessorOutputKind::TextileIntermediate),
        "weaver" => Some(ProcessorOutputKind::Cloth),
        "charcoal_burner" => Some(ProcessorOutputKind::Charcoal),
        "smithy" => Some(ProcessorOutputKind::Ironwork),
        "potter_kiln" => Some(ProcessorOutputKind::Pottery),
        "tannery" => Some(ProcessorOutputKind::Leather),
        "cobbler" => Some(ProcessorOutputKind::Shoes),
        "chandlery" => Some(ProcessorOutputKind::Candles),
        _ => None,
    }
}

pub fn processor_input_kinds(kind: &str) -> &'static [ProcessorInputKind] {
    use ProcessorInputKind::*;

    match kind {
        "watermill" | "windmill" => &[Grain],
        "bakery" => &[Flour, Water, Firewood],
        "brewery" => &[Barley, Water, Firewood, Apples, Honey],
        "smokehouse" => &[Food, Firewood, Salt],
        "spinning_retting_house" => &[Wool, Flax, Water],
        "weaver" => &[Yarn, Linen],
        "charcoal_burner" => &[Firewood],
        "smithy" => &[Iron, Charcoal, Water],
        "potter_kiln" => &[Clay, Firewood, Water],
        "tannery" => &[Hides, Water, Firewood],
        "cobbler" => &[Leather],
        "chandlery" => &[Wax, Firewood],
        _ => &[],
    }
}

pub fn is_processor_output_target_kind(kind: &str) -> bool {
    processor_output_kind(kind).is_some()
}

pub fn normalize_processor_output_target_percent(_percent: u8) -> u8 {
    PROCESSOR_OUTPUT_TARGET_DEFAULT_PERCENT
}

pub fn processor_output_target(capacity: f64, percent: u8) -> f64 {
    if !capacity.is_finite() {
        return 0.0;
    }
    capacity.max(0.0) * normalize_processor_output_target_percent(percent) as f64 / 100.0
}

pub fn processor_output_headroom(stock: f64, capacity: f64, percent: u8) -> f64 {
    if !stock.is_finite() {
        return 0.0;
    }
    (processor_output_target(capacity, percent) - stock.max(0.0)).max(0.0)
}

pub fn processor_input_staging_cycles(_percent: u8) -> f64 {
    PROCESSOR_INPUT_STAGING_DEFAULT_CYCLES
}

#[cfg(test)]
mod tests {
    use super::{
        is_processor_output_target_kind, normalize_processor_output_target_percent,
        processor_input_kinds, processor_input_staging_cycles, processor_output_headroom,
        processor_output_kind, processor_output_target, ProcessorInputKind, ProcessorOutputKind,
        PROCESSOR_INPUT_STAGING_DEFAULT_CYCLES, PROCESSOR_OUTPUT_TARGET_DEFAULT_PERCENT,
    };

    #[test]
    fn legacy_processors_keep_filling_to_capacity() {
        assert_eq!(PROCESSOR_OUTPUT_TARGET_DEFAULT_PERCENT, 100);
        assert_eq!(normalize_processor_output_target_percent(0), 100);
        assert_eq!(processor_output_target(240.0, 100), 240.0);
        assert_eq!(processor_output_headroom(80.0, 240.0, 0), 160.0);
    }

    #[test]
    fn legacy_percentages_all_resolve_to_physical_capacity() {
        for percent in [0, 25, 50, 75, 100, 255] {
            assert_eq!(normalize_processor_output_target_percent(percent), 100);
        }
    }

    #[test]
    fn production_uses_physical_capacity() {
        assert_eq!(processor_output_target(240.0, 25), 240.0);
        assert_eq!(processor_output_headroom(45.0, 240.0, 25), 195.0);
        assert_eq!(processor_output_headroom(240.0, 240.0, 25), 0.0);
    }

    #[test]
    fn workshops_stage_three_input_cycles_automatically() {
        assert_eq!(processor_input_staging_cycles(25), 3.0);
        assert_eq!(processor_input_staging_cycles(50), 3.0);
        assert_eq!(processor_input_staging_cycles(75), 3.0);
        assert_eq!(processor_input_staging_cycles(100), 3.0);
        assert_eq!(processor_input_staging_cycles(0), 3.0);
        assert_eq!(PROCESSOR_INPUT_STAGING_DEFAULT_CYCLES, 3.0);
    }

    #[test]
    fn only_staffed_conversion_workshops_use_automatic_capacity() {
        for kind in [
            "watermill",
            "windmill",
            "bakery",
            "brewery",
            "smokehouse",
            "spinning_retting_house",
            "weaver",
            "charcoal_burner",
            "smithy",
            "potter_kiln",
            "chandlery",
        ] {
            assert!(is_processor_output_target_kind(kind));
        }
        for kind in ["monastery", "apiary", "carpenter"] {
            assert!(!is_processor_output_target_kind(kind));
        }
        assert_eq!(
            processor_output_kind("watermill"),
            Some(ProcessorOutputKind::Flour)
        );
        assert_eq!(
            processor_output_kind("windmill"),
            Some(ProcessorOutputKind::Flour)
        );
        assert_eq!(
            processor_output_kind("smokehouse"),
            Some(ProcessorOutputKind::PreservedFood)
        );
    }

    #[test]
    fn geological_worksites_do_not_use_workshop_output_capacity() {
        for kind in ["stone_quarry", "large_quarry", "mine"] {
            assert!(!is_processor_output_target_kind(kind));
        }
        assert!(!is_processor_output_target_kind("pastoral_farmstead"));
    }

    #[test]
    fn every_processor_exposes_its_complete_authoritative_input_recipe() {
        use ProcessorInputKind::*;

        assert_eq!(processor_input_kinds("watermill"), &[Grain]);
        assert_eq!(processor_input_kinds("windmill"), &[Grain]);
        assert_eq!(processor_input_kinds("bakery"), &[Flour, Water, Firewood]);
        assert!(processor_input_kinds("granary").is_empty());
        assert_eq!(
            processor_input_kinds("brewery"),
            &[Barley, Water, Firewood, Apples, Honey]
        );
        assert_eq!(
            processor_input_kinds("smokehouse"),
            &[Food, Firewood, Salt]
        );
        assert_eq!(
            processor_input_kinds("spinning_retting_house"),
            &[Wool, Flax, Water]
        );
        assert_eq!(processor_input_kinds("weaver"), &[Yarn, Linen]);
        assert_eq!(processor_input_kinds("charcoal_burner"), &[Firewood]);
        assert_eq!(processor_input_kinds("smithy"), &[Iron, Charcoal, Water]);
        assert_eq!(
            processor_input_kinds("potter_kiln"),
            &[Clay, Firewood, Water]
        );
        assert_eq!(processor_input_kinds("chandlery"), &[Wax, Firewood]);
        assert_eq!(
            processor_output_kind("chandlery"),
            Some(ProcessorOutputKind::Candles)
        );
    }

    #[test]
    fn malformed_stock_or_capacity_cannot_create_headroom() {
        assert_eq!(processor_output_target(f64::NAN, 100), 0.0);
        assert_eq!(processor_output_target(-10.0, 100), 0.0);
        assert_eq!(processor_output_headroom(f64::NAN, 240.0, 100), 0.0);
    }
}

File: src/economy/processorOutputPolicy.ts
import { BUILDING_STORAGE_CAPS } from '../generated/gameBalance.ts';
import type { BuildingKind, BuildingState } from '../resources/types.ts';
import { preservedFoodStock } from './foodInventory.ts';
import { breadStock, flourStock } from './cropGoods.ts';
import {
  normalizePotterFiringPolicy,
  POTTER_FIRE_ROOF_TILES,
} from './potterFiringPolicy.ts';
import {
  breweryPolicyOutput,
  breweryRecipeRequestsInput,
  type BreweryRecipeInput,
} from './breweryRecipePolicy.ts';
import {
  selectedSmokehouseRecipePolicy,
  smokehouseRecipeOutput,
} from './smokehouseRecipePolicy.ts';
import {
  spinningRettingRecipeRequestsInput,
  weaverRecipeRequestsInput,
  weaverUsesFlax,
} from './weaverInputPolicy.ts';
import {
  isStorageCommodity,
  storageAcceptsCommodity,
} from './storageAcceptancePolicy.ts';
import { buildingSharedStorageRoom } from './sharedStorageCapacity.ts';

export const PROCESSOR_OUTPUT_TARGET_KINDS = [
  'watermill',
  'windmill',
  'bakery',
  'brewery',
  'smokehouse',
  'spinning_retting_house',
  'weaver',
  'charcoal_burner',
  'smithy',
  'potter_kiln',
  'tannery',
  'cobbler',
  'chandlery',
] as const satisfies readonly BuildingKind[];

export type ProcessorOutputTargetKind =
  (typeof PROCESSOR_OUTPUT_TARGET_KINDS)[number];

export const EXTRACTION_OUTPUT_TARGET_KINDS = [
  'stone_quarry',
  'large_quarry',
  'mine',
] as const satisfies readonly BuildingKind[];

export type ExtractionOutputTargetKind =
  (typeof EXTRACTION_OUTPUT_TARGET_KINDS)[number];

export type ProcessorOutputCommodity =
  | 'flour'
  | 'bread'
  | 'ale'
  | 'cider'
  | 'pearCider'
  | 'mead'
  | 'curedMeat'
  | 'smokedFish'
  | 'cheese'
  | 'yarn'
  | 'linen'
  | 'cloth'
  | 'charcoal'
  | 'ironwork'
  | 'pottery'
  | 'roofTiles'
  | 'leather'
  | 'shoes'
  | 'candles';

export type ExtractionOutputCommodity = 'stone' | 'iron' | 'salt' | 'clay';

export type ProcessorInputCommodity =
  | 'ryeGrain'
  | 'oatGrain'
  | 'maslinGrain'
  | 'barley'
  | 'malt'
  | 'apples'
  | 'pears'
  | 'honey'
  | 'ryeFlour'
  | 'maslinFlour'
  | 'water'
  | 'firewood'
  | 'food'
  | 'wool'
  | 'flax'
  | 'yarn'
  | 'linen'
  | 'iron'
  | 'clay'
  | 'salt'
  | 'charcoal'
  | 'pottery'
  | 'ironwork'
  | 'hides'
  | 'leather'
  | 'wax';

export const PROCESSOR_OUTPUT_TARGET_DEFAULT_PERCENT = 100;
export const PROCESSOR_INPUT_STAGING_DEFAULT_CYCLES = 3;

const OUTPUT_BY_KIND: Record<
  ProcessorOutputTargetKind,
  ProcessorOutputCommodity
> = {
  watermill: 'flour',
  windmill: 'flour',
  bakery: 'bread',
  brewery: 'ale',
  smokehouse: 'curedMeat',
  spinning_retting_house: 'yarn',
  weaver: 'cloth',
  charcoal_burner: 'charcoal',
  smithy: 'ironwork',
  potter_kiln: 'pottery',
  tannery: 'leather',
  cobbler: 'shoes',
  chandlery: 'candles',
};

const INPUTS_BY_KIND: Record<
  ProcessorOutputTargetKind,
  readonly ProcessorInputCommodity[]
> = {
  watermill: ['ryeGrain', 'maslinGrain'],
  windmill: ['ryeGrain', 'maslinGrain'],
  bakery: ['ryeFlour', 'maslinFlour', 'water', 'firewood'],
  brewery: ['barley', 'malt', 'apples', 'pears', 'honey', 'water', 'firewood'],
  smokehouse: ['food', 'firewood', 'salt'],
  spinning_retting_house: ['wool', 'flax', 'water'],
  weaver: ['yarn', 'linen'],
  charcoal_burner: ['firewood'],
  smithy: ['iron', 'charcoal', 'water'],
  potter_kiln: ['clay', 'firewood', 'water'],
  tannery: ['hides', 'water', 'firewood'],
  cobbler: ['leather'],
  chandlery: ['wax', 'firewood'],
};

export function isProcessorOutputTargetKind(
  kind: BuildingKind,
): kind is ProcessorOutputTargetKind {
  return (PROCESSOR_OUTPUT_TARGET_KINDS as readonly BuildingKind[]).includes(kind);
}

export function isExtractionOutputTargetKind(
  kind: BuildingKind,
): kind is ExtractionOutputTargetKind {
  return (EXTRACTION_OUTPUT_TARGET_KINDS as readonly BuildingKind[]).includes(kind);
}

export function normalizeProcessorOutputTargetPercent(
  _percent: number | undefined,
): number {
  return PROCESSOR_OUTPUT_TARGET_DEFAULT_PERCENT;
}

export function processorOutputCommodity(
  kind: ProcessorOutputTargetKind,
): ProcessorOutputCommodity {
  return OUTPUT_BY_KIND[kind];
}

export function processorOutputCommodityForBuilding(
  building: Pick<
    BuildingState,
    | 'kind'
    | 'potterFiringPolicy'
    | 'breweryRecipePolicy'
    | 'smokehouseRecipePolicy'
    | 'barley'
    | 'malt'
    | 'apples'
    | 'honey'
    | 'meat'
    | 'fish'
    | 'milk'
    | 'wool'
    | 'flax'
    | 'water'
    | 'weaverInputPolicy'
  >,
): ProcessorOutputCommodity | null {
  if (!isProcessorOutputTargetKind(building.kind)) return null;
  if (building.kind === 'brewery') {
    return breweryPolicyOutput(building.breweryRecipePolicy, building);
  }
  if (building.kind === 'smokehouse') {
    return smokehouseRecipeOutput(
      selectedSmokehouseRecipePolicy(building.smokehouseRecipePolicy, building),
    );
  }
  if (building.kind === 'spinning_retting_house') {
    return weaverUsesFlax(building) ? 'linen' : 'yarn';
  }
  if (
    building.kind === 'potter_kiln'
    && normalizePotterFiringPolicy(building.potterFiringPolicy) === POTTER_FIRE_ROOF_TILES
  ) {
    return 'roofTiles';
  }
  return processorOutputCommodity(building.kind);
}

export function processorInputCommodities(
  kind: ProcessorOutputTargetKind,
): readonly ProcessorInputCommodity[] {
  return INPUTS_BY_KIND[kind];
}

export function processorOutputTarget(
  capacity: number,
  percent: number | undefined,
): number {
  if (!Number.isFinite(capacity)) return 0;
  return Math.max(0, capacity)
    * normalizeProcessorOutputTargetPercent(percent)
    / 100;
}

export function processorInputStagingCycles(
  _percent: number | undefined,
): number {
  return PROCESSOR_INPUT_STAGING_DEFAULT_CYCLES;
}

export function processorOutputTargetForBuilding(
  building: Pick<
    BuildingState,
    | 'kind'
    | 'processorOutputTargetPercent'
    | 'potterFiringPolicy'
    | 'breweryRecipePolicy'
    | 'smokehouseRecipePolicy'
    | 'barley'
    | 'malt'
    | 'apples'
    | 'honey'
    | 'meat'
    | 'fish'
    | 'milk'
    | 'wool'
    | 'flax'
    | 'water'
    | 'weaverInputPolicy'
  >,
): number | null {
  const output = processorOutputCommodityForBuilding(building);
  if (!output || !isProcessorOutputTargetKind(building.kind)) return null;
  const capacities = BUILDING_STORAGE_CAPS[building.kind] as Partial<
    Record<ProcessorOutputCommodity | 'food' | 'preservedFood', number>
  >;
  const capacity = output === 'bread'
    ? capacities.food ?? 0
    : output === 'curedMeat' || output === 'smokedFish' || output === 'cheese'
      ? capacities.preservedFood ?? 0
    : capacities[output] ?? 0;
  return processorOutputTarget(capacity, building.processorOutputTargetPercent);
}

export function processorOutputHeadroom(
  building: BuildingState,
): number | null {
  const output = processorOutputCommodityForBuilding(building);
  if (!output) return null;
  const target = processorOutputTargetForBuilding(building) ?? 0;
  const stock = building.kind === 'smokehouse'
    ? preservedFoodStock(building)
    : output === 'flour'
      ? flourStock(building)
      : output === 'bread'
        ? breadStock(building)
        : Math.max(0, building[output] ?? 0);
  return Math.max(0, target - stock);
}

export function extractionOutputTarget(
  kind: ExtractionOutputTargetKind,
  commodity: ExtractionOutputCommodity,
): number {
  return (
    BUILDING_STORAGE_CAPS[kind] as Partial<
      Record<ExtractionOutputCommodity, number>
    >
  )[commodity] ?? 0;
}

export function extractionOutputHeadroom(
  building: BuildingState,
  commodity: ExtractionOutputCommodity,
): number | null {
  if (!isExtractionOutputTargetKind(building.kind)) return null;
  const target = extractionOutputTarget(
    building.kind,
    commodity,
  );
  return Math.min(
    Math.max(0, target - Math.max(0, building[commodity] ?? 0)),
    buildingSharedStorageRoom(building),
  );
}

export function extractionOutputCommodity(
  kind: BuildingKind,
  mineralResource: 'iron' | 'salt' | 'clay' | null = null,
): ExtractionOutputCommodity | null {
  switch (kind) {
    case 'stone_quarry':
    case 'large_quarry':
      return 'stone';
    case 'mine':
      return mineralResource;
    default:
      return null;
  }
}

/**
 * Replacement tools are durable worksite reserves, but they still need room in
 * the same physical yard. Unknown mineral sites remain ineligible because they
 * have no valid workface.
 */
export function extractionAcceptsMaintenance(
  building: BuildingState,
  mineralResource: 'iron' | 'salt' | 'clay' | null = null,
): boolean {
  if (!isExtractionOutputTargetKind(building.kind)) return true;
  const output = extractionOutputCommodity(building.kind, mineralResource);
  return output !== null && buildingSharedStorageRoom(building) > 1e-6;
}

export function processorNeedsInputs(
  building: BuildingState,
): boolean {
  const headroom = processorOutputHeadroom(building);
  return headroom == null || headroom > 1e-6;
}

export function processorUsesInput(
  kind: BuildingKind,
  commodity: ProcessorInputCommodity,
): boolean {
  return isProcessorOutputTargetKind(kind)
    && INPUTS_BY_KIND[kind].includes(commodity);
}

/**
 * New logistics demand is narrower than physical acceptance for a focused
 * recipe. Existing alternate stock and already-inbound cargo remain valid.
 */
export function processorRequestsInput(
  building: Pick<BuildingState, 'kind' | 'breweryRecipePolicy' | 'weaverInputPolicy'>,
  commodity: ProcessorInputCommodity,
): boolean {
  if (building.kind === 'brewery') {
    return !processorUsesInput('brewery', commodity)
      || breweryRecipeRequestsInput(
        building.breweryRecipePolicy,
        commodity as BreweryRecipeInput,
      );
  }
  if (building.kind === 'spinning_retting_house') {
    return commodity !== 'wool' && commodity !== 'flax' && commodity !== 'water'
      || spinningRettingRecipeRequestsInput(building.weaverInputPolicy, commodity);
  }
  if (building.kind === 'weaver') {
    return commodity !== 'yarn' && commodity !== 'linen'
      || weaverRecipeRequestsInput(building.weaverInputPolicy, commodity);
  }
  // Smokehouse food demand is typed at the dispatch source; both kiln recipes
  // share all inputs, so neither needs filtering here.
  return true;
}

export function processorAcceptsInput(
  building: BuildingState,
  commodity: ProcessorInputCommodity,
): boolean {
  if (isStorageCommodity(commodity) && !storageAcceptsCommodity(building, commodity)) {
    return false;
  }
  if (
    (building.kind === 'watermill' || building.kind === 'windmill')
    && commodity === 'oatGrain'
  ) return false;
  if (
    building.kind === 'pastoral_farmstead'
    && (
      commodity === 'ryeGrain'
      || commodity === 'oatGrain'
      || commodity === 'maslinGrain'
    )
  ) {
    return commodity === 'oatGrain'
      && (building.animalFeed ?? 0) + 1e-6
        < (BUILDING_STORAGE_CAPS.pastoral_farmstead.animalFeed ?? 0);
  }
  if (building.kind === 'pastoral_farmstead' && commodity === 'salt') {
    return preservedFoodStock(building) + 1e-6
      < (BUILDING_STORAGE_CAPS.pastoral_farmstead.preservedFood ?? 0);
  }
  if (
    building.kind === 'spinning_retting_house'
    && (commodity === 'wool' || commodity === 'flax' || commodity === 'water')
  ) {
    const output = commodity === 'wool' ? 'yarn' : 'linen';
    const capacity = BUILDING_STORAGE_CAPS.spinning_retting_house[output] ?? 0;
    const target = processorOutputTarget(capacity, building.processorOutputTargetPercent);
    return Math.max(0, building[output] ?? 0) + 1e-6 < target;
  }
  return !processorUsesInput(building.kind, commodity)
    || processorNeedsInputs(building);
}
